onnxruntime/core/providers/shared_library/provider_interfaces.h

```diff
@@ -562,6 +562,7 @@ struct ProviderHost {
   virtual const std::vector<MLDataType>& DataTypeImpl__AllTensorTypes() = 0;
   virtual const std::vector<MLDataType>& DataTypeImpl__AllIEEEFloatTensorTypes() = 0;
   virtual const std::vector<MLDataType>& DataTypeImpl__AllTensorAndSequenceTensorTypes() = 0;
+  virtual const std::vector<MLDataType>& DataTypeImpl__AllOptionalAndTensorAndSequenceTensorTypes() = 0;
```

Comment:
Do we need to introduce this in the shared provider files given that there are no EPs other than the CPU EP that implements these kernels for now ? (You will notice that `AllOptionalTypes()` haven't made it into these files yet)